### PR TITLE
Fixes #25921 Sync Capsules before applying errata at incremental update

### PR DESCRIPTION
**Pull request by pmoravec**
Identify smart proxies the Hosts are registered through, and invoke Capsule
synchronization of the updated content before applying errata to those Hosts.

This will make some of the automatically triggered Capsule synchronizations
redundant, but such no-op is harmless and preventing it could be error-prone.

Fixes: #25921

Signed-off-by: Pavel Moravec <pmoravec@redhat.com>

### CODE REVIEW

**Comment by pmoravec** [comment]
I think the test `if ::Katello::CapsuleContent.sync_needed?(env)` can be redundant as if no sync is needed for the environment, `::Katello::CapsuleContent.with_environment(env)` will be empty and `if capsules.any?` won't trigger the Capsule sync against empty Capsules.

But a confirmation from somebody else would be appreciated.

**Comment by pmoravec** [comment]
Current failure is some CI issue (postgres `database "test-katello-pr-test-4510-production" does not exist`), other tests passed.

**Comment by jlsherrill** [comment]
[test katello]

**Comment by bkearney** [comment]
@jlsherrill bump on this, I am looking to close out some older bugs.

File: app/lib/actions/katello/content_view/incremental_updates.rb [hunk 1]
```diff
@@ -29,14 +35,40 @@ def plan(version_environments, composite_version_environments, content, dep_solv
...
+            smart_proxies = (smart_proxies.map { |proxy| SmartProxy.find_by_name(proxy) unless proxy.nil? || SmartProxy.find_by_name(proxy).default_capsule? }).compact
+          end
+          return hosts, smart_proxies
+        end
+
+        def plan_smart_proxy_sync(smart_proxies, environment, version)
+          if smart_proxies.any?
+            concurrence do
+              environment.each do |env|
+                if ::Katello::CapsuleContent.sync_needed?(env)
```
Comment (jlsherrill) [inline app/lib/actions/katello/content_view/incremental_updates.rb]: CapsuleContent no longer exists in master, removed in: https://github.com/Katello/katello/pull/7750

Moved to the SmartProxy class

Comment (jlsherrill) [reply app/lib/actions/katello/content_view/incremental_updates.rb]: but i think you are correct, this is redundant to the ::Katello::CapsuleContent.with_environment(env).map { |capsule| capsule.capsule }

File: app/lib/actions/katello/content_view/incremental_updates.rb [hunk 2]
```diff
@@ -29,14 +35,40 @@ def plan(version_environments, composite_version_environments, content, dep_solv
...
             plan_self(:version_outputs => output_for_version_ids)
           end
         end
 
+        def find_applicable_hosts_and_smart_proxies(hosts, content)
+          smart_proxies = []
+          if hosts.any? && !content[:errata_ids].blank?
+            errata = ::Katello::Erratum.with_identifiers(content[:errata_ids])
+            hosts = hosts.where(:id => ::Katello::Host::ContentFacet.with_applicable_errata(errata).pluck(:host_id))
+            smart_proxies = (hosts.map { |host| host.subscription_facet_attributes.registered_through unless host.subscription_facet_attributes.nil? }).uniq
```
Comment (jlsherrill) [inline app/lib/actions/katello/content_view/incremental_updates.rb]: something like:
```
Katello::Host::SubscriptionFacet.where(:host => hosts).pluck(:registered_through).uniq.compact
```

File: app/lib/actions/katello/content_view/incremental_updates.rb [hunk 3]
```diff
@@ -29,14 +35,40 @@ def plan(version_environments, composite_version_environments, content, dep_solv
...
+            smart_proxies = (smart_proxies.map { |proxy| SmartProxy.find_by_name(proxy) unless proxy.nil? || SmartProxy.find_by_name(proxy).default_capsule? }).compact
```
Comment (jlsherrill) [inline app/lib/actions/katello/content_view/incremental_updates.rb]: This could be re-written as:  
```
smart_proxies = SmartProxy.where(:name => smart_proxies)
```
I'm not sure we need to include the default capsule here?

File: app/lib/actions/katello/content_view/incremental_updates.rb (hunk 1 above)
Comment (pmoravec) [reply app/lib/actions/katello/content_view/incremental_updates.rb]: Thanks for confirmation, removing it.

File: app/lib/actions/katello/content_view/incremental_updates.rb [hunk 4]
```diff
@@ -29,14 +35,40 @@ def plan(version_environments, composite_version_environments, content, dep_solv
...
+          end
+          return hosts, smart_proxies
+        end
+
+        def plan_smart_proxy_sync(smart_proxies, environment, version)
+          if smart_proxies.any?
+            concurrence do
+              environment.each do |env|
+                if ::Katello::CapsuleContent.sync_needed?(env)
+                  capsules = ::Katello::CapsuleContent.with_environment(env).map { |capsule| capsule.capsule }
```
[the remark this answers is not in the document]
Comment (pmoravec) [reply app/lib/actions/katello/content_view/incremental_updates.rb]: Renaming.. :)

File: app/lib/actions/katello/content_view/incremental_updates.rb (hunk 2 above)
Comment (pmoravec) [reply app/lib/actions/katello/content_view/incremental_updates.rb]: Updating for more ruby-like style :)

File: app/lib/actions/katello/content_view/incremental_updates.rb (hunk 3 above)
Comment (pmoravec) [reply app/lib/actions/katello/content_view/incremental_updates.rb]: Good point; while `hosts` must contain also hosts under default Capsule (and they do), `smart_proxies` shouldnt contain the default one.

**Comment by pmoravec** [comment]
Rebased to current master, comments incorporated. I am not fully happy with finding nondefault proxies based on their name:

    smart_proxies = SmartProxy.where(:name => smart_proxies).where.not(:id => SmartProxy.default_capsule!.id)

that might have a simplier solution with just one `where` condition - but I havent figured it out how..

**Comment by jturel** [comment]
[test katello]

File: app/lib/actions/katello/content_view/incremental_updates.rb [hunk 5]
```diff
@@ -29,14 +35,38 @@ def plan(version_environments, composite_version_environments, content, dep_solv
...
             end
             plan_self(:version_outputs => output_for_version_ids)
           end
         end
 
+        def find_applicable_hosts_and_smart_proxies(hosts, content)
+          smart_proxies = []
+          if hosts.any? && !content[:errata_ids].blank?
+            errata = ::Katello::Erratum.with_identifiers(content[:errata_ids])
+            hosts = hosts.where(:id => ::Katello::Host::ContentFacet.with_applicable_errata(errata).pluck(:host_id))
```
Comment (jturel) [inline app/lib/actions/katello/content_view/incremental_updates.rb]: Hey @pmoravec I'm trying to understand why you added a second lookup/filter for hosts when the action has already been told which hosts to be concerned with. It appears to me that the controller which invokes this action has already determined the applicable hosts here: https://github.com/Katello/katello/blob/a1c9280067607999cae43bab89b53ba870856b76/app/controllers/katello/api/v2/content_view_versions_controller.rb#L162-L176

In other words: I have a suspicion that the `.where` here does not actually do any extra filtering and would return the same set of hosts

Comment (jturel) [reply app/lib/actions/katello/content_view/incremental_updates.rb]: I see the code was written that way before, and it's probably just a safeguard

File: app/lib/actions/katello/content_view/incremental_updates.rb [hunk 6]
```diff
@@ -29,14 +35,38 @@ def plan(version_environments, composite_version_environments, content, dep_solv
...
-              hosts = hosts.where(:id => ::Katello::Host::ContentFacet.with_applicable_errata(errata).pluck(:host_id))
               plan_action(::Actions::BulkAction, ::Actions::Katello::Host::Erratum::ApplicableErrataInstall, hosts, content[:errata_ids])
             end
             plan_self(:version_outputs => output_for_version_ids)
           end
         end
 
+        def find_applicable_hosts_and_smart_proxies(hosts, content)
+          smart_proxies = []
+          if hosts.any? && !content[:errata_ids].blank?
```
Comment (jturel) [inline app/lib/actions/katello/content_view/incremental_updates.rb]: Here's a possible refactor of this block of code:

```
-          if hosts.any? && !content[:errata_ids].blank?
+          if hosts.any? && content[:errata_ids].present?
             errata = ::Katello::Erratum.with_identifiers(content[:errata_ids])
-            hosts = hosts.where(:id => ::Katello::Host::ContentFacet.with_applicable_errata(errata).pluck(:host_id))
-            smart_proxies = Katello::Host::SubscriptionFacet.where(:host => hosts).pluck(:registered_through).uniq.compact
-            smart_proxies = SmartProxy.where(:name => smart_proxies).where.not(:id => SmartProxy.default_capsule!.id)
+            facets = ::Katello::Host::ContentFacet.with_applicable_errata(errata)
+            hosts = hosts.where(id: facets.pluck(:host_id))
+            smart_proxies = SmartProxy.non_default.where(content_facets: facets)
```
```
+++ b/app/models/katello/concerns/smart_proxy_extensions.rb
@@ -55,6 +55,7 @@ module Katello
           :message => _("must be one of the following: %s") % DOWNLOAD_POLICIES.join(', ')
         }
         scope :with_content, -> { with_features(PULP_FEATURE, PULP_NODE_FEATURE) }
+        scope :non_default, -> { with_features(PULP_NODE_FEATURE) }
```

I haven't tested it yet but it should solve the problem of looking up the proxies by name

**Comment by jturel** [comment]
@pmoravec ping :)

File: app/lib/actions/katello/content_view/incremental_updates.rb (hunk 6 above)
Comment (pmoravec) [reply app/lib/actions/katello/content_view/incremental_updates.rb]: Sounds nice, will improve this way.

File: app/lib/actions/katello/content_view/incremental_updates.rb (hunk 5 above)
Comment (pmoravec) [reply app/lib/actions/katello/content_view/incremental_updates.rb]: Indeed, I just moved that filter from original line 33 here / where it is applicable in new code. But it really seems it is a redundant filter as you pointed out, `calculate_hosts_for_incremental` seems to be executed everytime before this.

I am leaving the filter as a safe-guard now as a conservative/minimalistic change - if you think it is really safe to remove it, I can include the removal in my PR.

**Comment by pmoravec** [comment]
Suggestions to code improvement applied, the safe-guarded code left there as it was there before.

**Comment by pmoravec** [comment]
Tests failed on `test_distribution_repositories` which does not seem to be relevant to this PR but rather some temporary postgres issue..? (`User.current = User.find(users(:admin).id)` call failed)

File: app/lib/actions/katello/content_view/incremental_updates.rb [hunk 7]
```diff
@@ -7,36 +7,66 @@ class IncrementalUpdates < Actions::EntryAction
...
+            smart_proxies = SmartProxy.non_default.where(content_facets: facets)
+          end
+          return hosts, smart_proxies
+        end
+
+        def plan_smart_proxy_sync(smart_proxies, environment, version)
+          if smart_proxies.any?
+            concurrence do
+              environment.each do |env|
+                capsules = ::SmartProxy.with_environment(env).map { |capsule| capsule.capsule }
```
Comment (jturel) [inline app/lib/actions/katello/content_view/incremental_updates.rb]: I think this can be rewritten from:
```
capsules = ::SmartProxy.with_environment(env).map { |capsule| capsule.capsule }
capsules &= smart_proxies
```

to

```
environment_proxies = smart_proxies.with_environment(env)
```
why?
- `capsule.capsule` no longer works on master branch
- slightly more clear variable naming
- I think we get the same result (let me know if I missed something)

**Comment by jturel** [comment]
Here's some rough draft of what tests I think we should have in test/actions/katello/content_view_test.rb#IncrementalUpdatesTest
```
   it 'plans capsule syncs when needed' do
      Dynflow::Testing::DummyPlannedAction.any_instance.stubs(:new_content_view_version).returns(::Katello::ContentViewVersion.first)

      plan_action(action, [{:content_view_version => content_view.version(library), :environments => [library]}], [],
                  {:errata_ids => ["FOO"]}, true, [], "BadDescription")

      expected_capsules = []
      expected_cv_id = nil
      expected_env_id = nil

      assert_action_planed_with(action, ::Actions::BulkAction, expected_capsules, content_view_id: expected_cv_id, environment_id: expected_env_id)
    end

    it 'plans errata install when needed' do

    end
```

**Comment by pmoravec** [comment]
Updated to the `smart_proxies.with_environment(env)` (I usually prepare and test a patch on older `katello` that might divert from upstream, so some my constructions are not up-to-date :) ).

For the tests: good point though I have no experience in writing or testing them. I would need to spend more time I dont have now, so I just applied the first improvement.

**Comment by pmoravec** [comment]
About tests where I am pure newbie and write code without testing it: here is what the two test cases shall look like (I bet there are multiple issues so havent re-spin the PR):

    --- test/actions/katello/content_view_test.rb	2019-04-05 08:41:09.775789157 +0200
    +++ content_view_test.rb	2019-05-30 13:17:29.095426303 +0200
    @@ -353,6 +353,48 @@ module ::Actions::Katello::ContentView
           end
         end
     
    +    it 'plans capsule syncs when needed' do
    +      Dynflow::Testing::DummyPlannedAction.any_instance.stubs(:new_content_view_version).returns(::Katello::ContentViewVersion.first)
    +
    +      smart_proxy_service_1 = new_capsule_content(:five)
    +      smart_proxy_service_2 = new_capsule_content(:six)
    +      smart_proxy_service_1.smart_proxy.add_lifecycle_environment(library)  # just test one proxy will get the content
    +
    +      plan_action(action, [{:content_view_version => content_view.version(library), :environments => [library]}], [],
    +                  {:errata_ids => ["FOO"]}, true, [], "BadDescription")
    +
    +      assert_action_planned_with(action, ::Actions::BulkAction, ::Actions::Katello::CapsuleContent::Sync,
    +                                 [smart_proxy_service_1.smart_proxy],
    +                                 :content_view_id => content_view.id, :environment_id => library.id)
    +    end
    +
    +    it 'plans errata install when needed' do
    +      Dynflow::Testing::DummyPlannedAction.any_instance.stubs(:new_content_view_version).returns(::Katello::ContentViewVersion.first)
    +
    +      proxy = FactoryBot.build(:smart_proxy, :url => 'http://fakepath.com/foo')
    +      smart_proxy_service = capsule_content(:seven)
    +      smart_proxy_service.smart_proxy.add_lifecycle_environment(library)
    +      proxy.capsule_content = smart_proxy_service    # ??? how to pair proxy to capsule content?
    +      proxy.save!
    +
    +      errata = katello_errata("bugfix")
    +
    +      host = FactoryBot.create(:host, :with_content, :content_view => @view,
    +                                          :lifecycle_environment => @library)
    +      host.content_facet.content_source = proxy
    +      host.content_facet.applicable_errata << errata
    +      host.save!
    +
    +      plan_action(action, [{:content_view_version => content_view.version(library), :environments => [library]}], [],
    +                  {:errata_ids => [errata.errata_id]}, true, [], "BadDescription")
    +
    +      assert_action_planned_with(action, ::Actions::BulkAction, Actions::Katello::Host::Erratum::ApplicableErrataInstall, [host], [errata.errata_id]).returns({})  # not sure with the "returns", it was copied from elsewhere
    +
    +      put :install_content, params: { :included => {:ids => [@host1.id]}, :organization_id => @org.id, :content_type => 'errata', :content => [errata.errata_id] } # ??? copied from errata apply test, not sure if applicable - but we need to verify the errata will be installed on a host registered to the Capsule
    +      assert_response :success # ???
    +
    +    end
    +
         it 'generates correct message' do
           data = {:version_outputs =>
                    [{:version_id => 1,

**Comment by jturel** [comment]
I'll take a shot at the tests so we can get this contribution merged in. Will get back in a day or two!

**Comment by pmoravec** [comment]
Thanks for collaboration.

I merged the 3 commits into one and rebased the PR.

**Comment by jturel** [comment]
@jlsherrill you've looked at this PR in the past - would you mind giving it a once-over for the approach taken since pavel and I have worked together on it a bit?

**Comment by jturel** [comment]
@pmoravec I was giving this a final test-through and discovered that the query to look up the proxies which I pointed you to will not work as intended. I'll spare the details unless you want me to explain.

@jlsherrill suggested to me that we don't try to isolate which proxies need syncing by which hosts are registered through them, but rather to **sync _all_ the proxies which have the affected environment** and then run the applicable errata install as before. It's possible we'll sync a few extra proxies, but we'll still guarantee that the affected hosts can install those errata.

It also occurred to us that this change could greatly increase the time it takes to perform the incremental update since the capsule sync becomes synchronous to overall execution as it's written now. We think it's a good idea to effectively mirror what is being done during CV promote in order to **run capsule sync and applicable errata install asynchronous to the incremental update** Example code is here: https://github.com/Katello/katello/blob/3329a4f0336f37db3bd026b93c96904b012391f1/app/lib/actions/katello/content_view/promote_to_environment.rb#L57-L60
I think we could create a new action to handle this async behavior such as Actions::Katello::ContentView::SyncAndInstallErrata (just a suggested name)

I've emphasized in bold the main ideas of the changes suggested here - let me know if I've explained them well and if you agree with them. If you're not sure how to approach making the changes I can offer more guidance as needed :)

**Comment by pmoravec** [comment]
I agree with the proposal of changes. Due to the holiday time and some other commitment, I can get to this in a month manner the soonest.

**Comment by bkearney** [comment]
@pmoravec @jturel what is the status of this pr?

**Comment by jturel** [comment]
This needs a rebase and I believe there was some feedback pending code changes

**Comment by jturel** [comment]
@pmoravec any chance you'll be able to get back to this? Do you need us to take it over? (I see there's a BZ for it)

**Comment by pmoravec** [comment]
Sorry, no time to rework the patch :( A take over would be appreciated.

**Comment by jturel** [comment]
Thanks Pavel I'll work on it this sprint

**Comment by jturel** [comment]
Closing this in favor of #8719 - thanks Pavel!
